### PR TITLE
xrCore build on Linux

### PR DESCRIPTION
**Pull request by eagleivg**
xrCore now building under Linux, implemented work with the file system, determine the capabilities of the processor, the number of occupied and free memory, event management. To check up working capacity while it is not obviously possible before end of porting of other parts of the project.

### CODE REVIEW

File: src/xrCore/LocatorAPI_defs.cpp [hunk 1]
```diff
@@ -4,22 +4,30 @@
...
 FS_File::FS_File(const xr_string& nm, long sz, time_t modif, unsigned attr) { set(nm, sz, modif, attr); }
 FS_File::FS_File(const xr_string& nm) { set(nm, 0, 0, 0); }
+#if defined(WINDOWS)
 FS_File::FS_File(const _FINDDATA_T& f) { set(f.name, f.size, f.time_write, (f.attrib & _A_SUBDIR) ? flSubDir : 0); }
 FS_File::FS_File(const xr_string& nm, const _FINDDATA_T& f)
 {
     set(nm, f.size, f.time_write, (f.attrib & _A_SUBDIR) ? flSubDir : 0);
 }
+#else
+FS_File::FS_File(const _FINDDATA_T& f) { ; }
```
Comment (Xottab-DUTY) [inline src/xrCore/LocatorAPI_defs.cpp]: Following the code style, this should be `{}` instead of `{ ; }`

File: src/xrCore/ModuleLookup.cpp [hunk 2]
```diff
@@ -52,6 +53,8 @@ void* ModuleHandle::operator()() const
 
 void* ModuleHandle::getProcAddress(pcstr procName) const
 {
+#if defined(WINDOWS)
     return GetProcAddress(static_cast<HMODULE>(handle), procName);
+#endif
```
Comment (Xottab-DUTY) [inline src/xrCore/ModuleLookup.cpp:58]: Function without a return. :D
Anyway, I will upload ModuleHandle Linux implementation soon.

File: src/xrCore/Threading/Event.cpp [hunk 3]
```diff
@@ -1,12 +1,81 @@
 #include "stdafx.h"
 #include "Event.hpp"
+#if defined(WINDOWS)
+#include <windows.h>
```
Comment (Xottab-DUTY) [inline src/xrCore/Threading/Event.cpp]: I don't see a reason to include <windows.h> anywhere. It's already included in global header `Common/PlatformWindows.hpp`

File: src/xrCore/Threading/Event.hpp [hunk 4]
```diff
@@ -3,7 +3,20 @@
...
+	struct EventHandle
+	{
+		pthread_mutex_t mutex;
+		pthread_cond_t cond;
+		bool signaled;
+	};
+    pthread_mutex_t handle;
+
+private:
+	EventHandle m_id;
```
Comment (Xottab-DUTY) [inline src/xrCore/Threading/Event.hpp]: Tabs should be replaced with spaces (1 tab -> 4 spaces)

File: src/xrCore/lzo_compressor.h [hunk 5]
```diff
@@ -1,12 +0,0 @@
-#pragma once
```
Comment (Xottab-DUTY) [inline src/xrCore/lzo_compressor.h:1]: It was unused?

File: src/xrCore/xrDebug.cpp [hunk 6]
```diff
@@ -360,6 +390,7 @@ void xrDebug::Fail(bool& ignoreAlways, const ErrorLocation& loc, const char* exp
         OnDialog(true);
     FlushLog();
     if (Core.PluginMode)
+#if defined(WINDOWS)
```
Comment (Xottab-DUTY) [inline src/xrCore/xrDebug.cpp]: For Linux you have this now:
```cpp
if (Core.PluginMode)
    lock.Leave();
```

File: src/xrCore/xrDebug.cpp [hunk 7]
```diff
@@ -391,6 +422,7 @@ void xrDebug::Fail(bool& ignoreAlways, const ErrorLocation& loc, const char* exp
     }
     if (OnDialog)
```
Comment (Xottab-DUTY) [inline src/xrCore/xrDebug.cpp]: `OnDialog` should not be hidden by `#if defined(WINDOWS)` =)

File: src/xrCore/xrMemory.h [hunk 8]
```diff
@@ -11,7 +11,9 @@
 Поэтому всё-же стоит переопределять для большинства случаев операторы new и delete.
 А для остального мы будем полагать (и надеяться), что прокси справится без проблем.
 */
+#if defined(WINDOWS) // Не знаю, как это на виндах работает, но на линуксе ломается линковка из-за множественного объявления __TBB_malloc_proxy_helper_object
```
Comment (Xottab-DUTY) [inline src/xrCore/xrMemory.h]: Плохо, что ломается =(((

Впрочем, TBB итак подцепляется через перегрузки операторов new и собственные функции.

File: src/xrCore/xrMemory.cpp [hunk 9]
```diff
@@ -77,8 +100,11 @@ void xrMemory::mem_compact()
         g_pStringContainer->clean();
     if (g_pSharedMemoryContainer)
         g_pSharedMemoryContainer->clean();
-    if (strstr(Core.Params, "-swap_on_compact"))
+
+#if defined(WINDOWS)
+if (strstr(Core.Params, "-swap_on_compact"))
```
Comment (Xottab-DUTY) [inline src/xrCore/xrMemory.cpp]: Случайно отступ удалился)

File: src/xrCore/xrstring.h [hunk 10]
```diff
@@ -199,10 +199,30 @@ IC void xr_strlwr(shared_str& src)
     if (*src)
     {
         char* lp = xr_strdup(*src);
+#if defined(WINDOWS)
         xr_strlwr(lp);
+#elif defined(LINUX)
+        int i = 0;
+        while(lp[i])
+        {
+        	lp[i] = (char) toupper(lp[i]);// TODO rewrite locale-independent toupper_l()
```
Comment (Xottab-DUTY) [inline src/xrCore/xrstring.h]: Maybe we should use `std` functions that depend on `std::locale`?

**Review by Xottab-DUTY — COMMENTED** [review]
This is great!

File: src/xrCore/LocatorAPI.cpp [hunk 11]
```diff
@@ -642,6 +670,7 @@ bool CLocatorAPI::Recurse(pcstr path)
     // insert self
     if (path && path[0] != 0)
         Register(path, 0xffffffff, 0, 0, 0, 0, 0);
+
```
Comment (Kaffeine) [inline src/xrCore/LocatorAPI.cpp:673]: Unrelated formatting change

Comment (Xottab-DUTY) [reply src/xrCore/LocatorAPI.cpp:673]: I think it's fine =)

File: src/xrCore/xrMemory.h (hunk 8 above)
Comment (Kaffeine) [reply src/xrCore/xrMemory.h]: IIRC only english comments are acceptable in OpenXRay development. Please rewrite it to something like this:

```diff
-#if defined(WINDOWS) // Не знаю, как это на виндах работает, но на линуксе ломается линковка из-за множественного объявления __TBB_malloc_proxy_helper_object
+#if defined(WINDOWS) // I have not idea how it works on Windows, but Linux build fails with error 'multiple declaration of __TBB_malloc_proxy_helper_object'
```

Better yet: remove the comment and extract the change to a commit "Fix multiple declaration of '__TBB_malloc_proxy_helper_object' error on Linux"

Better yet: properly fix the problem for all platforms.

File: src/xrCore/xr_ini.cpp [hunk 12]
```diff
@@ -6,6 +6,207 @@
...
+            if(cur<'a' || cur>='a'+base-10)
+                break;
+            v = cur-'a'+10;
+        }
+
+        nptr++;
+
+        if(ret>MSVCRT_UI64_MAX/base || ret*base>MSVCRT_UI64_MAX-v) {
+            ret = MSVCRT_UI64_MAX;
+        } else
```
Comment (Kaffeine) [inline src/xrCore/xr_ini.cpp]: The style looks odd. Please either use { } for both one-line branches or for none.

File: src/xrCore/lzo_compressor.h (hunk 5 above)
Comment (eagleivg) [reply src/xrCore/lzo_compressor.h:1]: xrCore/lzo_compressor.* is duplicate of xrCore/Compression/lzo_compressor.*

File: src/xrCore/xr_ini.cpp [hunk 13]
```diff
@@ -6,6 +6,207 @@
...
+#define MSVCRT_UI64_MAX   (((uint64_t)0xffffffff << 32) | 0xffffffff)
+
+/**
+ * from wine@dlls/msvcrt/string.c
+ *
+ * @param fileName
+ * @param readOnly
+ * @return
+ */
+int _cdecl _i64toa_s(int64_t value, char *str, size_t size, int radix){
```
Comment (Kaffeine) [inline src/xrCore/xr_ini.cpp]: The change set does not follow coding style.

**Comment by Kaffeine** [comment]
Nice job! Thank you so much for the effort!

File: src/xrCore/lzo_compressor.h (hunk 5 above)
Comment (Xottab-DUTY) [reply src/xrCore/lzo_compressor.h:1]: Ah, okay.

File: src/xrCore/xrMemory.h (hunk 8 above)
Comment (Xottab-DUTY) [reply src/xrCore/xrMemory.h]: > Для Windows прокси линкуется со всем, что можно. Visual Studio это переваривает.
В принципе прокси можно линковать только для основного бинарника. Он будет прекрасно работать.
И я думаю это исправит проблему с gcc.

>For Windows TBB proxy is linked to everything. Visual Studio accepts that.
Well, we can link it to main executable only. It would work excellent.
And I think this would fix problem with GCC.

As suggested by @intorr, we can link TBB proxy only to main executable

**Comment by Xottab-DUTY** [comment]
@Kaffeine, are we okay to merge? Issue with TBB can be fixed in the next PR.

**Comment by Kaffeine** [comment]
@Xottab-DUTY I'm not in the position to approve or decline the change, but yes, this looks good to me. 
IMHO it would be better to have this merged.
